File: frontend/types/quest.ts
// 펫·레벨·데일리 퀘스트 도메인 타입(PRD.md §14.5). 계약: docs/quest-pet-api.md.
export interface PetState {
  code: string;
  name: string;
  emoji: string;
  stage_label: string;
}

/** 고를 수 있는 펫(카탈로그는 서버가 준다 — FE가 따로 들지 않는다). */
export interface PetOption {
  code: string;
  name: string;
  emoji: string;
}

export interface QuestState {
  code: string;
  label: string;
  exp: number;
  is_done: boolean;
}

export interface QuestProgress {
  level: number;
  exp: number;
  exp_into_level: number;
  exp_per_level: number;
  pet: PetState;
  pets: PetOption[];
  quests: QuestState[];
}

/** 퀘스트 코드 — 발화 지점이 화면 곳곳이라 문자열 오타를 타입으로 막는다. */
export const QUEST = {
  viewShort: "view_short",
  viewLong: "view_long",
  askChat: "ask_chat",
} as const;

File: backend/app/schemas/quest.py
from pydantic import BaseModel, Field


class PetState(BaseModel):
    code: str
    name: str
    emoji: str  # 현재 레벨 단계의 외형(ponytail: 임시 — 일러스트 나오면 이미지 URL로 교체)
    stage_label: str


class PetOption(BaseModel):
    """고를 수 있는 펫. emoji는 **현재 레벨 기준** 외형이라 "지금 바꾸면 이렇게 보인다"가 된다."""

    code: str
    name: str
    emoji: str


class QuestState(BaseModel):
    code: str
    label: str
    exp: int
    is_done: bool


class QuestProgress(BaseModel):
    """펫·레벨·오늘 퀘스트 = 화면 하나를 채우는 상태 전부(PRD.md §14.5).
    조회/완료/펫교체 셋 다 이걸 돌려줘 FE가 응답을 통째로 갈아끼우면 되게 한다."""

    level: int
    exp: int
    exp_into_level: int  # 현재 레벨 안에서 쌓은 양(진행바)
    exp_per_level: int
    pet: PetState
    pets: list[PetOption]
    quests: list[QuestState]


class PetChangeRequest(BaseModel):
    code: str = Field(min_length=1, max_length=32)  # 실제 유효성은 서비스가 화이트리스트로 검증
